File: src/algorithms/python/Largest_Rectangle_in_Histogram.py
# -*- coding: utf-8 -*-


class Solution:
    # @param {integer[]} height
    # @return {integer}
    def largestRectangleArea(self, height):
        st, ret = [], 0
        height.append(0)
        for i in range(len(height)):
            if st == []:
                st.append((height[i], i))
            else:
                if height[i] >= st[-1][0]:
                    st.append((height[i], i))
                else:
                    while st != [] and height[i] < st[-1][0]:
                        top = st.pop()
                        pre = -1 if st == [] else st[-1][1]
                        w, h = i - pre - 1, top[0]
                        ret = max(ret, w * h)
                    st.append((height[i], i))
        return ret


if __name__ == '__main__':
    solution = Solution()
    print solution.largestRectangleArea([2, 1, 5, 6, 2, 3])
    print solution.largestRectangleArea([2, 1, 2])
    print solution.largestRectangleArea([2, 0, 2])

File: src/algorithms/cpp/Largest_Rectangle_in_Histogram.cpp
#include <iostream>
#include <cstdio>
#include <cstring>
#include <cmath>
#include <cstdlib>
#include <algorithm>
#include <vector>
#include <map>
#include <set>
#include <stack>
using namespace std;
#define out(v) cerr << #v << ": " << (v) << endl
#define SZ(v) ((int)(v).size())
const int maxint = -1u>>1;
template <class T> bool get_max(T& a, const T &b) {return b > a? a = b, 1: 0;}
template <class T> bool get_min(T& a, const T &b) {return b < a? a = b, 1: 0;}

class Solution {
public:
    int largestRectangleArea(vector <int>& height) {
        stack <pair<int, int> > st;
        int result = 0;
        height.push_back(0);
        for (int i = 0; i < SZ(height); ++ i) {
            while (!st.empty() && height[i] < st.top().first) {
                pair <int, int> top = st.top();
                st.pop();
                int pre;
                if (st.empty())     pre = -1;
                else                pre = st.top().second;
                int w = i - pre - 1, h = top.first;
                get_max(result, w * h);
            }
            st.push(make_pair(height[i], i));
        }
        return result;
    }
};

int main() {
    vector <int> input = {2, 1, 5, 6, 2, 3};
    Solution solution = Solution();
    cout << solution.largestRectangleArea(input) << endl;
    return 0;
}
